### PR TITLE
Add bootsnap to optimize boot time

### PR DESCRIPTION
**Pull request by Watson1978**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Description

My previous pull-request (https://github.com/fastlane/fastlane/pull/10536) has a problem in dependency gem (bootsnap) for cli tools.
Finally, new version of bootsnap would be released and solved the problem.

--

Currentry, `fastlane` takes a time nearly 1 sec even if it displays help message.

```
$ time bundle exec fastlane -h > /dev/null
bundle exec fastlane -h > /dev/null  0.73s user 0.22s system 98% cpu 0.966 total
```

If you create new lane/action by trial and error, I think it is useful if boot time will be reduced.
This patch will reduce the boot time 0.966 sec -> 0.584 sec using `bootsnap` gem.
https://github.com/Shopify/bootsnap

Seems `fastlane` requires many Ruby files such as plugins/helpers and it takes a time where parses/generates AST and Ruby instruction sequences.

`bootsnap` gem generates Ruby instruction sequences cache and it will be able to reduce boot time.

At first boot, `bootsnap` generates Ruby instruction sequences cache and it take a time.

```
$ time bundle exec fastlane -h > /dev/null
bundle exec fastlane -h > /dev/null  0.94s user 0.45s system 98% cpu 1.420 total
```

But, after generated Ruby instruction sequences cache, boot time will be reduced.

```
$ time bundle exec fastlane -h > /dev/null
bundle exec fastlane -h > /dev/null  0.43s user 0.14s system 97% cpu 0.584 total
```

My environment is
* MacBook 2017
* CPU : 1.4 GHz Core i7
* Memory : 16 GB

### CODE REVIEW

**Review by giginet — DISMISSED** [review]
Thank you for your PR 😄 

https://github.com/fastlane/fastlane/blob/master/CORE_CONTRIBUTOR.md

According to our guideline, external dependencies should be keeping to minimum.

I have no idea this change how affects impact to boot time.
So I think this patch should not be merged 🙏

**Comment by taquitos** [comment]
I agree with @giginet, unfortunately, we have many different ways that fastlane is packaged (homebrew, Fabric Mac App package, gem, source), and we've seen some unexpected behavior when we start using optimizations like this.

**Comment by KrauseFx** [comment]
Yah, that's a tough one. Thanks for your work here @Watson1978, I wonder if it's easier and safer to optimize when and how we require other Ruby files, instead of adding a dependency to make the other things faster?

**Comment by janpio** [comment]
From my viewpoint a 50% boot time reduction would definitely be worth the added dependency. This gem is maintained by Shopify and also used by Discourse, so I think we don't have to fear abandonment or something like that.

**Comment by KrauseFx** [comment]
Agreed - if we can make things faster without breaking stuff, that would be 💯

**Comment by janpio** [comment]
So this becomes the question how we can make sure this doesn't break fastlane with all "different ways that fastlane is packaged (homebrew, Fabric Mac App package, gem, source),"...

**Comment by Watson1978** [comment]
I think bootsnap is good gem and many people think so, and then bootsnap will be introduced as default gem at Rails 5.2.0 :)

 http://weblog.rubyonrails.org/2017/11/27/Rails-5-2-Active-Storage-Redis-Cache-Store-HTTP2-Early-Hints-Credentials/

**Comment by giginet** [comment]
In my opinion, I'm afraid this patch causing side effects.
However, we never know before merging.
Trying this is low risk. We can come back anytime by just reverting this PR.

So I leave to core developers' choices.

**Comment by janpio** [comment]
@Watson1978 Just checked which platforms bootsnap supports. Could you maybe checkout the branch from #11242, cherry-pick your changes on there and submit this as another, `[WIP]`, PR so the tests can be run on Ubuntu as well? Would really improve my confidence in adding this.

**Comment by janpio** [comment]
Actually seems that Windows support was added as well: https://github.com/Shopify/bootsnap/pull/60
That might get handy soon... #windows

**Comment by lacostej** [comment]
What we could for things we are afraid of introducing breakages is to implement a simple mechanism to do A/B testing and feature flags.

Only make those experimental features for people who opt-in experimental setups. And recruit people.

**Review by janpio — COMMENTED** [review]
One required change for Windows compat.

**Review by janpio — CHANGES_REQUESTED** [review]
see above

**Comment by janpio** [comment]
I resolved the conflict, but this also merged the newest CI configs and general code changes which really don't seem to like bootsnap:

```
[12:30:22]: fastlane finished with errors
bundler: failed to load command: fastlane (/Users/distiller/project/.bundle/ruby/2.3.0/bin/fastlane)
LoadError: cannot load such file -- ../fastlane/lib/fastlane/version.rb
  /Users/distiller/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:19:in `require'
  Fastfile:434:in `block in parsing_binding'
  /Users/distiller/project/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/distiller/project/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'
  /Users/distiller/project/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'
  /Users/distiller/project/fastlane/lib/fastlane/fast_file.rb:148:in `method_missing'
  Fastfile:47:in `block in parsing_binding'
  /Users/distiller/project/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/distiller/project/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
  /Users/distiller/project/fastlane/lib/fastlane/runner.rb:45:in `chdir'
  /Users/distiller/project/fastlane/lib/fastlane/runner.rb:45:in `execute'
  /Users/distiller/project/fastlane/lib/fastlane/lane_manager.rb:56:in `cruise_lane'
  /Users/distiller/project/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
  /Users/distiller/project/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
  /Users/distiller/project/.bundle/ruby/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
  /Users/distiller/project/.bundle/ruby/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
  /Users/distiller/project/.bundle/ruby/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
  /Users/distiller/project/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:75:in `run!'
  /Users/distiller/project/.bundle/ruby/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
  /Users/distiller/project/fastlane/lib/fastlane/commands_generator.rb:337:in `run'
  /Users/distiller/project/fastlane/lib/fastlane/commands_generator.rb:42:in `start'
  /Users/distiller/project/fastlane/lib/fastlane/cli_tools_distributor.rb:98:in `take_off'
  /Users/distiller/project/bin/fastlane:34:in `<top (required)>'
  /Users/distiller/project/.bundle/ruby/2.3.0/bin/fastlane:23:in `load'
  /Users/distiller/project/.bundle/ruby/2.3.0/bin/fastlane:23:in `<top (required)>'
```

and

```
bundler: failed to load command: fastlane (/home/circleci/project/.bundle/ruby/2.3.0/bin/fastlane)
ArgumentError: Trying to register Bundler::GemfileError for status code 4 but Bundler::GemfileError is already registered
  /usr/local/lib/ruby/gems/2.3.0/gems/bundler-1.16.1/lib/bundler/errors.rb:9:in `status_code'
  /usr/local/lib/ruby/gems/2.3.0/gems/bundler-1.16.1/lib/bundler/errors.rb:20:in `<class:GemfileError>'
  /usr/local/lib/ruby/gems/2.3.0/gems/bundler-1.16.1/lib/bundler/errors.rb:20:in `<module:Bundler>'
  /usr/local/lib/ruby/gems/2.3.0/gems/bundler-1.16.1/lib/bundler/errors.rb:3:in `<main>'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /usr/local/lib/ruby/gems/2.3.0/gems/bundler-1.16.1/lib/bundler.rb:10:in `<main>'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /home/circleci/project/fastlane/lib/fastlane/plugins/plugin_manager.rb:5:in `<class:PluginManager>'
  /home/circleci/project/fastlane/lib/fastlane/plugins/plugin_manager.rb:4:in `<module:Fastlane>'
  /home/circleci/project/fastlane/lib/fastlane/plugins/plugin_manager.rb:3:in `<main>'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /home/circleci/project/fastlane/lib/fastlane/plugins/plugins.rb:9:in `<main>'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /home/circleci/project/fastlane/lib/fastlane.rb:27:in `<main>'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /home/circleci/project/.bundle/ruby/2.3.0/gems/bootsnap-1.1.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
  /home/circleci/project/fastlane/lib/fastlane/cli_tools_distributor.rb:37:in `take_off'
  /home/circleci/project/bin/fastlane:34:in `<top (required)>'
  /home/circleci/project/.bundle/ruby/2.3.0/bin/fastlane:23:in `load'
  /home/circleci/project/.bundle/ruby/2.3.0/bin/fastlane:23:in `<top (required)>'
```

:/

**Comment by janpio** [comment]
In general I agree @lacostej re feature flags, but how would you implement that for `bin/fastlane`?

File: bin/fastlane
Comment (Watson1978) [reply bin/fastlane]: Fixed at https://github.com/fastlane/fastlane/pull/11230/commits/8b787e0f2da2f719affe1c468883a658e6c46e8e

**Comment by janpio** [comment]
I tried this again at https://github.com/fastlane/fastlane/pull/13167 building on the startup measurement I built in https://github.com/fastlane/fastlane/pull/13166. The whole thing activated does work locally on my Windows machine, but not on CI at all :/

**Pull request closed without merging**
